Lib/fontTools/cffLib/specializer.py

```diff
@@ -26,26 +27,62 @@ def programToString(program):
 	return ' '.join(str(x) for x in program)
 
 
-def programToCommands(program):
+def programToCommands(program, deltasToSourceValues=None, numRegions=0):
 	"""Takes a T2CharString program list and returns list of commands.
 	Each command is a two-tuple of commandname,arg-list.  The commandname might
 	be empty string if no commandname shall be emitted (used for glyph width,
 	hintmask/cntrmask argument, as well as stray arguments at the end of the
-	program (¯\_(ツ)_/¯)."""
+	program. The latter happens in subroutines.
+	Note that each 'blend' operator and its argument list is  replaced by a
+	list of arguments in which each argument item is a list of source font
+	values.
+	'deltasToSourceValues' can be from varLib.varStore.VarDataItem."""
 
 	width = None
 	commands = []
 	stack = []
+	seen_blend = False
+	seen_op = False
+	num_sources = numRegions + 1
+
 	it = iter(program)
 	for token in it:
 		if not isinstance(token, basestring):
 			stack.append(token)
 			continue
-
-		if width is None and token in {'hstem', 'hstemhm', 'vstem', 'vstemhm',
-					       'cntrmask', 'hintmask',
-					       'hmoveto', 'vmoveto', 'rmoveto',
-					       'endchar'}:
+		if token == 'blend':
+			# if we see a blend, it is a CFF2 charstring, and there
+			# is never a width arg.
+			if not seen_op:
+				seen_op = True
+			num_blends = stack.pop()
+			if not seen_blend:
+				seen_blend = True
+				assert deltasToSourceValues is not None, (
+					"Cannot process charstring without varDataItem argument")
+
+			num_args = num_sources * num_blends
+			argi = len(stack) - num_args
+			# argi is the index of first blend arg, aka first default font value
+			# We will replace this with a list for the source font values.
+			end_args = tuplei = argi + num_blends
+			while argi < end_args:
+				next_ti = tuplei + numRegions
+				deltas = stack[tuplei:next_ti]
+				default_value = stack[argi]
+				blend_list = deltasToSourceValues(default_value, deltas)
+				stack[argi] = blend_list
+				tuplei = next_ti
+				argi += 1
+			del stack[-(num_args- num_blends):]
+			# convert blend op to blend value list args
+			continue
+			
+		if not seen_op and token in {'hstem', 'hstemhm', 'vstem', 'vstemhm',
```

Comment:
This code attempts to process the width in a CFF2 CharString, if no `blend` op has been seen so far and an extra argument is on the stack. It shouldn't be triggered by a well-formed CFF2 CharString, but seems odd and maybe better to explicitly check for CFF2 instead?

Reply:
The 'is _CFF2' state would have to be passed in as a function argument: it is not currently available in the function. This would certainly be a simpler solution.